### PR TITLE
CIP-25 validation + version 2

### PR DESCRIPTION
**Pull request by vladimirvolek**
- Proper validation for CIP-25 on-chain metadata
- Support for CIP-25 version 2
- Preview integration tests + CI
- Increase test timeouts
- toBe -> toSrictEqual
- Integration tests fixes

resolves: https://github.com/blockfrost/blockfrost-backend-ryo/issues/13

### CODE REVIEW

File: src/routes/addresses/index.ts [hunk 1]
```diff
@@ -131,13 +135,24 @@ async function addresses(fastify: FastifyInstance) {
 
         if (rows[0].amount) {
           for (const asset of rows[0].amount) {
-            const registryData = await fetchAssetMetadata(asset.unit);
+            const unit = `${asset.policy_id}${asset.asset_name}`;
```
Comment (slowbackspace) [inline src/routes/addresses/index.ts:138]: `asset.unit` is no longer available?

File: src/utils/validation.ts [hunk 2]
```diff
@@ -31,3 +35,48 @@ export const isTestnet = (): boolean => {
...
+    const onchainMetadataResult = onchainMetadata[policyId][assetName];
+    const isSchemaValid = validateOnchainMetadata(onchainMetadataResult);
+
+    if (!isSchemaValid) {
+      return null;
+    }
+
+    return onchainMetadataResult;
+  } catch (error) {
+    console.log(error);
```
[the remark this answers is not in the document]
Comment (slowbackspace) [reply src/utils/validation.ts:79]: otherwise we won't know from where it is firing

File: src/utils/validation.ts [hunk 3]
```diff
@@ -31,3 +35,48 @@ export const isTestnet = (): boolean => {
...
+
+  return 1;
+};
+
+export const getOnchainMetadata = (
+  onchainMetadata: Asset['onchain_metadata'],
+  assetName: Asset['asset_name'],
+  policyId: Asset['policy_id'],
+) => {
+  if (!onchainMetadata || !assetName) return null;
```
Comment (slowbackspace) [inline src/utils/validation.ts:56]: couldn't assetName be an empty string? that's still a valid token (composed of only policy id) that could have some onchain metadata?

File: src/utils/validation.ts [hunk 4]
```diff
@@ -31,3 +35,48 @@ export const isTestnet = (): boolean => {
...
+  const version = getOnchainMetadataVersion(onchainMetadata);
+
+  if (version === 1) {
+    assetName = Buffer.from(assetName, 'hex').toString('utf8');
+  }
+
+  // version 2 is default
+
+  try {
+    const assetSchema = getSchemaForEndpoint('/assets/{asset}');
```
Comment (slowbackspace) [inline src/utils/validation.ts:67]: would be nice to call this only once and not with every function call

File: src/routes/addresses/index.ts (hunk 1 above)
Comment (1000101) [reply src/routes/addresses/index.ts:138]: That is correct - we can keep it in the underlying SQL query but still need both name and policy for the validation function. It has to be concatenated in either SQL or JS, I chose the latter.

File: src/utils/validation.ts [hunk 5]
```diff
@@ -31,3 +35,48 @@ export const isTestnet = (): boolean => {
...
+    const onchainMetadataSchema = assetSchema['response']['200']['properties']['onchain_metadata'];
+    const validateOnchainMetadata = ajv.compile(onchainMetadataSchema);
```
Comment (slowbackspace) [inline src/utils/validation.ts:69]: same for the scehma compilation, it will speed up things in the long run.

File: src/utils/validation.ts [hunk 6]
```diff
@@ -31,3 +35,48 @@ export const isTestnet = (): boolean => {
...
+  onchainMetadata: Asset['onchain_metadata'],
+  assetName: Asset['asset_name'],
+  policyId: Asset['policy_id'],
+) => {
+  if (!onchainMetadata || !assetName) return null;
+
+  const version = getOnchainMetadataVersion(onchainMetadata);
+
+  if (version === 1) {
+    assetName = Buffer.from(assetName, 'hex').toString('utf8');
```
Comment (slowbackspace) [inline src/utils/validation.ts:61]: nitpick, mutating function parameters is not very nice!

File: src/utils/validation.ts (hunk 3 above)
Comment (vladimirvolek) [reply src/utils/validation.ts:56]: Can be empty string but null?

File: src/utils/validation.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vladimirvolek) [reply src/utils/validation.ts]: Do we want to log every invalid on-chain metadata to sentry?

File: src/utils/validation.ts (hunk 3 above)
Comment (vladimirvolek) [reply src/utils/validation.ts]: https://github.com/blockfrost/blockfrost-backend-ryo/pull/18/commits/96b81f7edc81977932209e0e8bccd290aa272bb5#diff-f9d8e0bc7f251fefd3b4f00fe043bd526c470ae6131c26b40f5ceb8f4264e4f1R577

File: src/utils/validation.ts (hunk 4 above)
Comment (vladimirvolek) [reply src/utils/validation.ts]: yes!

File: src/utils/validation.ts (hunk 5 above)
Comment (vladimirvolek) [reply src/utils/validation.ts]: 👍

File: src/utils/validation.ts (hunk 6 above)
Comment (vladimirvolek) [reply src/utils/validation.ts]: 👍

File: src/utils/validation.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (slowbackspace) [reply src/utils/validation.ts]: metadata that don't match the schema wont throw it simply returns false
```
 if (!isSchemaValid) {
      return null;
    }
```

the only thing you need to make sure is that you is that these obj indexes exist onchainMetadata[policyId][assetNameConverted], that would throw.

then THE ADVANTAGE would be that we would have error report in sentry if something really weird and fatal happened, and we should know about it without actively scanning for logs! but anyway that's is not that important if we already have at least easy to grep console.log with customized msg.

**Review by slowbackspace — APPROVED** [review]

File: src/utils/validation.ts [hunk 7]
```diff
@@ -31,3 +39,46 @@ export const isTestnet = (): boolean => {
...
+export const getOnchainMetadataVersion = (onchainMetadata: Asset['onchain_metadata']): 1 | 2 => {
+  if (!onchainMetadata?.version) {
+    return 1;
+  }
+
+  if (Number(onchainMetadata.version) === 2) {
+    return 2;
+  }
+
+  return 1;
```
Comment (slowbackspace) [inline src/utils/validation.ts:52]: just a random thought, if cardano introduces v3 then we will treat it as v1, not sure if that's a good idea but idk really. just something to keep in mind.

**Pull request merged**